Bio/Align/__init__.py

```diff
@@ -224,6 +224,38 @@ def _get_per_column_annotations(self):
         doc="""Dictionary of per-letter-annotation for the sequence.""",
     )
 
+    def _get_mol_type(self):
+        values = {_.annotations.get("molecule_type", None) for _ in self}
+        if not values:
+            return None  # no records, no molecule type
+        elif len(values) == 1:
+            # Easy case, perfect consensus
+            return list(values)[0]
+        elif all(_ and "DNA" in _ for _ in values):
+            return "DNA"  # e.g. mix of "DNA" and "gDNA"
+        elif all(_ and "RNA" in _ for _ in values):
+            return "RNA"  # e.g. mix of "RNA" and "mRNA"
+        elif all(_ and "DNA" in _ for _ in values):
+            return "protein"
+        else:
+            return None
+
+    def _set_mol_type(self, value):
+        for record in self:
+            record.annotations["molecule_type"] = value
```

Comment:
This seems odd. Is it needed?

Reply:
My approach here is rather than setting an MSA level attribute which can get out of sync, the property value is always computed on the fly from the record consensus.

In order to be able to set an alignment to "DNA" (for example), we applying that to all the records in the alignment. Making the user update/set this in all the records is tedious. Maybe I should make that more explicit in the unit tests...

My intension is to use this in the ``__init__`` method and/or when using the AlignIO parsing functions with an explicit molecule type.

Reply:
Yes but I am wondering why the molecule type of the records get changed when the molecule type of the alignment is changed.

Reply:
The multiple sequence alignment object is literally a collection of ``SeqRecord`` objects. By saying the alignment is DNA (or RNA or protein), you are by implication saying that all the records in it are also DNA. This does exactly that.

Reply:
To me that is an unexpected side effect. It will break in the future, and is inconsistent with how alphabets were used in `MultipleSeqAlignment`:
```
>>> from Bio.Alphabet import generic_dna
>>> from Bio.Alphabet import IUPAC
>>> from Bio.Seq import Seq
>>> from Bio.SeqRecord import SeqRecord
>>> from Bio.Align import MultipleSeqAlignment
>>> a = SeqRecord(Seq("AAAACGT", generic_dna), id="Alpha")
>>> b = SeqRecord(Seq("AAA-CGT", generic_dna), id="Beta")
>>> c = SeqRecord(Seq("AAAAGGT", generic_dna), id="Gamma")
>>> align = MultipleSeqAlignment([a, b, c], alphabet=IUPAC.unambiguous_dna)
```
Now the alignment alphabet is `IUPAC.unambiguous_dna`, but the individual sequences still have `generic_dna` as their alphabet:
```
>>> a.seq.alphabet
DNAAlphabet()
>>> a.seq.alphabet is generic_dna
True
>>> print(align)
IUPACUnambiguousDNA() alignment with 3 rows and 7 columns
AAAACGT Alpha
AAA-CGT Beta
AAAAGGT Gamma
```

Reply:
Yes, it is a change to the old alphabet behaviour, but it seemed easier to me to do it this way and ensure consistency, versus having separate potentially conflicting information at the alignment and record level. i.e. To me this is fixing a weakness of the old design. If we go with this, it can be more clearly emphasised in the documentation.

If you really don't like this approach, then how about a read only consensus property (i.e. remove the setter function)?

Or, would you prefer a simple attribute ``.molecule_type`` which is not directly linked to the record values (except perhaps via a default set during ``__init__``), like the current ``.alphabet`` attribute behaviour?

Reply:
It will cause problems for sure in the future. Suppose two alignments share some of the `SeqRecord`s. Then setting the molecule type in one alignment would set the molecule type of some of the `SeqRecord`s in the other alignment.

We should consider where the molecule type is actually needed. If it's only `AlignIO.write`, we could have an optional argument `molecule_type`. If `molecule_type` is not given, and the format needs to know the molecule type, then it can be derived from the `molecule_type` of the records as a default.

Reply:
If you have two alignments sharing records. and they are of mixed molecule types there certainly could be problems - but that's true regardless isn't it? A refinement of the proposal here would be an error/warning if changing a molecule type (e.g. RNA to DNA).

Code in ``AlignInfo`` also used the molecule type (via alphabet) for inferring expected letters in the sequence etc. See #2999 (which turned out to be more complicated that it first looked).

Reply:
> If you have two alignments sharing records. and they are of mixed molecule types there certainly could be problems - but that's true regardless isn't it?

No in general that is not a problem. For example, you could have an mRNA aligned to a chromosome sequences. Then when you use `AlignIO.write` to store the alignment in a format that requires a molecule type, and you set `align.molecule_type`, then suddenly your mRNA is genomic DNA or your chromosome is a mRNA.

> Code in AlignInfo also used the molecule type (via alphabet) for inferring expected letters in the sequence etc. See #2999 (which turned out to be more complicated that it first looked).

In `information_content`, you only need the expected letters to set the background frequency of each letter to 1 / number of letters. This is mathematically correct but biologically not very meaningful. So in any case you will need the background frequency as an argument to `information_content`. Then you won't need ot know the molecule type there.

Reply:
True, if the alignment tool was aware of what was going on for scoring, you could for DNA and RNA together - the pairwise example makes most sense.